File: backend/kernelCI_app/typeModels/metrics.py
from pydantic import BaseModel, Field

from kernelCI_app.constants.localization import DocStrings
from kernelCI_app.typeModels.metrics_notifications import (
    BuildIncidentsCount,
    LabMetricsData,
    MetricsReportData,
    TopIssue,
)

DEFAULT_METRICS_START_DAYS_AGO = 7
DEFAULT_METRICS_END_DAYS_AGO = 0


class MetricsQueryParameters(BaseModel):
    start_days_ago: int = Field(
        default=DEFAULT_METRICS_START_DAYS_AGO,
        ge=0,
        description=DocStrings.METRICS_START_DAYS_AGO_DESCRIPTION,
    )
    end_days_ago: int = Field(
        default=DEFAULT_METRICS_END_DAYS_AGO,
        ge=0,
        description=DocStrings.METRICS_END_DAYS_AGO_DESCRIPTION,
    )


class MetricsResponse(BaseModel):
    n_trees: int
    n_checkouts: int
    n_builds: int
    n_tests: int
    n_issues: int
    n_incidents: int
    build_incidents_by_origin: dict[str, BuildIncidentsCount]
    top_issues_by_origin: dict[str, list[TopIssue]]
    lab_maps: dict[str, LabMetricsData]
    prev_n_trees: int
    prev_n_checkouts: int
    prev_n_builds: int
    prev_n_tests: int
    prev_lab_maps: dict[str, LabMetricsData]


def metrics_report_data_to_response(data: MetricsReportData) -> MetricsResponse:
    return MetricsResponse(
        n_trees=data.n_trees,
        n_checkouts=data.n_checkouts,
        n_builds=data.n_builds,
        n_tests=data.n_tests,
        n_issues=data.n_issues,
        n_incidents=data.n_incidents,
        build_incidents_by_origin=data.build_incidents_by_origin,
        top_issues_by_origin={
            origin: list(issues.values())
            for origin, issues in data.top_issues_by_origin.items()
        },
        lab_maps=data.lab_maps,
        prev_n_trees=data.prev_n_trees,
        prev_n_checkouts=data.prev_n_checkouts,
        prev_n_builds=data.prev_n_builds,
        prev_n_tests=data.prev_n_tests,
        prev_lab_maps=data.prev_lab_maps,
    )

File: dashboard/src/types/metrics.ts
export type BuildIncidentsCount = {
  total_incidents: number;
  n_new_issues: number;
  n_existing_issues: number;
  n_total_issues: number;
};

export type TopIssue = {
  id: string;
  version: number;
  comment: string;
  total_incidents: number;
};

export type LabMetricsData = {
  builds: number;
  boots: number;
  tests: number;
};

export type MetricsResponse = {
  n_trees: number;
  n_checkouts: number;
  n_builds: number;
  n_tests: number;
  n_issues: number;
  n_incidents: number;
  build_incidents_by_origin: Record<string, BuildIncidentsCount>;
  top_issues_by_origin: Record<string, TopIssue[]>;
  new_issues_by_origin: Record<string, TopIssue[]>;
  lab_maps: Record<string, LabMetricsData>;
  prev_n_trees: number;
  prev_n_checkouts: number;
  prev_n_builds: number;
  prev_n_tests: number;
  prev_lab_maps: Record<string, LabMetricsData>;
};
